File: showcase/src/main/java/com/gigigo/showcase/ContentManager.java
package com.gigigo.showcase;

import android.app.Application;
import android.os.Handler;
import android.os.Looper;
import com.gigigo.orchextra.CrmUser;
import com.gigigo.orchextra.Orchextra;
import com.gigigo.orchextra.core.controller.model.home.ImageTransformReadArticle;
import com.gigigo.orchextra.ocm.OCManagerCallbacks;
import com.gigigo.orchextra.ocm.Ocm;
import com.gigigo.orchextra.ocm.OcmBuilder;
import com.gigigo.orchextra.ocm.OcmCallbacks;
import com.gigigo.orchextra.ocm.OcmEvent;
import com.gigigo.orchextra.ocm.OcmStyleUiBuilder;
import com.gigigo.orchextra.ocm.callbacks.OcmCredentialCallback;
import com.gigigo.orchextra.ocm.callbacks.OnEventCallback;
import com.gigigo.orchextra.ocm.views.UiGridBaseContentData;
import com.gigigo.showcase.main.MainActivity;
import com.gigigo.showcase.settings.ProjectData;
import com.gigigo.vuforiaimplementation.ImageRecognitionVuforiaImpl;
import java.util.GregorianCalendar;
import java.util.Map;

public class ContentManager {

  private static final ContentManager instance = new ContentManager();
  private static final String COUNTRY = "it";
  private Handler handler;

  public static ContentManager getInstance() {
    return instance;
  }

  private ContentManager() {
    this.handler = new Handler(Looper.getMainLooper());
  }

  public void init(Application application) {

    OcmBuilder ocmBuilder =
        new OcmBuilder(application).setNotificationActivityClass(MainActivity.class)
            .setShowReadArticles(true)
            .setTransformReadArticleMode(ImageTransformReadArticle.BITMAP_TRANSFORM)
            .setMaxReadArticles(100)
            .setOrchextraCredentials(ProjectData.getDefaultApiKey(),
                ProjectData.getDefaultApiSecret())
            .setContentLanguage("EN")
            .setOxSenderId("327008883283")
            .setOnEventCallback(new OnEventCallback() {
              @Override public void doEvent(OcmEvent event, Object data) {
              }

              @Override public void doEvent(OcmEvent event) {
              }
            });

    Ocm.initialize(ocmBuilder);

    OcmStyleUiBuilder ocmStyleUiBuilder =
        new OcmStyleUiBuilder().setTitleToolbarEnabled(true).setEnabledStatusBar(true);

    Ocm.setStyleUi(ocmStyleUiBuilder);
    Ocm.setBusinessUnit(COUNTRY);
  }

  public void start(ContentManagerCallback<String> callback) {
    start(ProjectData.getDefaultApiKey(), ProjectData.getDefaultApiSecret(), callback);
  }

  public void start(String apiKey, String apiSecret,
      final ContentManagerCallback<String> callback) {

    Ocm.setBusinessUnit(COUNTRY);
    Ocm.startWithCredentials(apiKey, apiSecret, new OcmCredentialCallback() {
      @Override public void onCredentialReceiver(final String accessToken) {
        handler.post(new Runnable() {
          @Override public void run() {
            callback.onSuccess(accessToken);
          }
        });
      }

      @Override public void onCredentailError(final String code) {
        handler.post(new Runnable() {
          @Override public void run() {
            callback.onError(new Exception(code));
          }
        });
      }
    });
  }

  public void clear() {
    Ocm.clearData(true, true, new OCManagerCallbacks.Clear() {
      @Override public void onDataClearedSuccessfull() {

      }

      @Override public void onDataClearFails(Exception e) {

      }
    });
  }

  public void getContent(String section, int imagesToDownload,
      final ContentManagerCallback<UiGridBaseContentData> callback) {

    Ocm.generateSectionView(section, null, imagesToDownload, new OcmCallbacks.Section() {
      @Override public void onSectionLoaded(UiGridBaseContentData uiGridBaseContentData) {
        callback.onSuccess(uiGridBaseContentData);
      }

      @Override public void onSectionFails(Exception e) {
        callback.onError(e);
      }
    });
  }

  public Map<String, String> getCustomFields() {
    return Orchextra.getUserCustomFields();
  }

  public void setUserCustomFields(Map<String, String> customFields) {
    Orchextra.bindUser(getCrmUser("test2"));
    Orchextra.setUserCustomFields(customFields);
    Orchextra.commitConfiguration();
  }

  private CrmUser getCrmUser(String id) {
    return new CrmUser(id, new GregorianCalendar(1990, 10, 15), CrmUser.Gender.GenderFemale);
  }

  public interface ContentManagerCallback<T> {

    void onSuccess(T result);

    void onError(Exception exception);
  }
}

File: showcase/src/main/java/com/gigigo/showcase/ocm/ContentManager.kt
package com.gigigo.showcase.ocm

import android.app.Application
import android.content.Intent
import android.net.Uri
import com.gigigo.orchextra.ocm.OCManagerCallbacks
import com.gigigo.orchextra.ocm.Ocm
import com.gigigo.orchextra.ocm.OcmBuilder
import com.gigigo.orchextra.ocm.OcmStyleUiBuilder
import com.gigigo.orchextra.ocm.callbacks.OcmCredentialCallback
import com.gigigo.orchextra.ocm.callbacks.OnCustomSchemeReceiver
import com.gigigo.orchextra.wrapper.CrmUser
import com.gigigo.orchextra.wrapper.OxManager
import com.gigigo.showcase.domain.entity.ConfigData
import com.gigigo.showcase.domain.entity.UserData
import com.gigigo.showcase.presentation.view.splash.SplashActivity
import timber.log.Timber
import java.util.HashMap

class ContentManager(private val context: Application) {

  private var token = ""

  fun init(configData: ConfigData, userData: UserData, onSuccess: () -> Unit = {},
      onError: () -> Unit = {}) {

    finish()

    val ocmBuilder = OcmBuilder(context).setOrchextraCredentials(configData.apiKey,
        configData.apiSecret)
        .setContentLanguage("en")
        .setBusinessUnit(configData.businessUnit)
        .setTriggeringEnabled(false)
        .setAnonymous(true)
        .setProximityEnabled(false)
        .setNotificationActivityClass(SplashActivity::class.java)

    Ocm.initialize(ocmBuilder, object : OcmCredentialCallback {
      override fun onCredentialReceiver(accessToken: String) {
        Timber.d("onCredentialReceiver")

        Ocm.setErrorListener(object : OxManager.ErrorListener {
          override fun onError(error: String) {
            Timber.e("Ox error: %s", error)
            onError()
          }
        })

        Ocm.setOnCustomSchemeReceiver(onCustomSchemeReceiver)

        val ocmStyleUiBuilder = OcmStyleUiBuilder()
            .setTitleToolbarEnabled(false)
            .disableThumbnailImages()
            .setEnabledStatusBar(false)
        Ocm.setStyleUi(ocmStyleUiBuilder)

        token = accessToken
//        bindUserData(userData)
        onSuccess()
      }

      override fun onCredentailError(code: String) {
        Timber.e("Error on init Ox: $code")
        if (!code.contains("tr.")) {
          onError()
        }
      }
    })
  }

  private fun bindUserData(userData: UserData, onSuccess: () -> Unit = {},
      onError: () -> Unit = {}) {

    Ocm.bindUser(CrmUser(userData.id, null, null), object : OxManager.StatusListener {
      override fun onSuccess() {
        val customFields = HashMap<String, String>()
        customFields["type"] = userData.type ?: ""
        customFields["level"] = userData.level ?: ""
        Ocm.setCustomFields(customFields, object : OxManager.StatusListener {
          override fun onSuccess() {
            onSuccess()
          }

          override fun onError(error: String) {
            Timber.e("setCustomFields error: %s", error)
            onError()
          }
        })
      }

      override fun onError(error: String) {
        Timber.e("bindUser error: %s", error)
        onError()
      }
    })
  }

  fun clearData(onSuccess: () -> Unit = {}, onError: () -> Unit = {}) {
    Ocm.clearData(true, true, object : OCManagerCallbacks.Clear {
      override fun onDataClearedSuccessfull() {
        onSuccess()
      }

      override fun onDataClearFails(e: Exception?) {
        Timber.e(e, "clearData()")
        onError()
      }
    })
  }

  private fun finish() {
    try {
      Ocm.stop()
    } catch (e: Exception) {
      Timber.e(e, "finish()")
    }
  }

  private val onCustomSchemeReceiver = OnCustomSchemeReceiver { scheme ->
    try {
      Timber.d("CustomScheme: %s", scheme)
      val intent = Intent()
      intent.action = "android.intent.action.VIEW"

      val deeplink = Uri.parse(scheme)
      intent.data = deeplink

      TODO("implement deeplinkHandler")
//      deeplinkHandler.processDeeplink(intent)

    } catch (exception: Exception) {
      Timber.e(exception, "onCustomSchemeReceiver")
    }
  }
}